File: web/src/lib/sessionWorkspace/subagents.ts
/**
 * Binding worker transcripts to the tool call that spawned them.
 *
 * Two shapes, two kinds of evidence, both provider-supplied:
 *
 * - A Task/Agent subagent names its parent tool call in a sidecar the engine
 *   reads at parse time, so the child arrives already carrying
 *   `parent_tool_call_id`. Exact match, nothing to infer.
 * - A Workflow subagent knows only its run id. The one place a run is tied to a
 *   tool call is the parent's own `Workflow` tool result, which names the
 *   transcript directory. That text is already in the transcript, so the
 *   binding is a read of evidence we have — not a heuristic.
 *
 * Fail closed on both. A run id that appears in no tool result leaves its
 * children unbound and the row renders as an ordinary tool call, which is what
 * it is until the provider says otherwise. Never "nearest Workflow call".
 */

import type { SubagentChild } from "./types";

/** `Transcript dir: …/subagents/workflows/<run>` — the literal line, nothing looser. */
const TRANSCRIPT_DIR_RUN = /Transcript dir:\s*\S*?[/\\]subagents[/\\]workflows[/\\]([A-Za-z0-9_-]+)/g;

/** Run ids named by one tool result, in the order they appear. */
export function runIdsInToolOutput(output: string | null | undefined): string[] {
  const text = output || "";
  if (!text.includes("subagents")) return [];
  const found: string[] = [];
  // A fresh regex per call: the shared literal carries lastIndex across uses.
  const pattern = new RegExp(TRANSCRIPT_DIR_RUN.source, "g");
  let match = pattern.exec(text);
  while (match !== null) {
    if (match[1] && !found.includes(match[1])) found.push(match[1]);
    match = pattern.exec(text);
  }
  return found;
}

/**
 * Children spawned by one tool call: those naming it directly, plus every
 * member of a run that call launched.
 */
export function childrenForToolCall(
  children: readonly SubagentChild[],
  { toolCallId, toolOutputText }: { toolCallId: string | null | undefined; toolOutputText: string | null | undefined },
): SubagentChild[] {
  if (children.length === 0) return [];
  const runIds = runIdsInToolOutput(toolOutputText);
  const matched = children.filter((child) => {
    if (toolCallId && child.parent_tool_call_id === toolCallId) return true;
    return Boolean(child.run_id) && runIds.includes(child.run_id as string);
  });
  return matched.sort((left, right) => (left.started_at || "").localeCompare(right.started_at || ""));
}

/** "22 agents · 4m12s" — the shape of the work, before anyone expands it. */
export function summarizeSubagents(children: readonly SubagentChild[]): string {
  if (children.length === 0) return "";
  const label = children.length === 1 ? "1 agent" : `${children.length} agents`;
  const starts = children.map((child) => child.started_at).filter(Boolean) as string[];
  const ends = children.map((child) => child.ended_at).filter(Boolean) as string[];
  if (starts.length === 0 || ends.length !== children.length) return label;
  const spanMs = Math.max(...ends.map((value) => Date.parse(value))) - Math.min(...starts.map((value) => Date.parse(value)));
  if (!Number.isFinite(spanMs) || spanMs <= 0) return label;
  const seconds = Math.round(spanMs / 1000);
  const duration = seconds < 60 ? `${seconds}s` : `${Math.floor(seconds / 60)}m${String(seconds % 60).padStart(2, "0")}s`;
  return `${label} · ${duration}`;
}

/** One child's line: its own title, or the prompt it was handed. */
export function subagentLabel(child: SubagentChild): string {
  const candidate = (child.title || child.first_user_message_preview || "").trim().replace(/\s+/g, " ");
  if (!candidate) return child.session_id.slice(0, 8);
  return candidate.length > 80 ? `${candidate.slice(0, 79)}…` : candidate;
}

File: ios/Sources/Shared/Subagents.swift
import Foundation

/// One worker transcript a tool call spawned. A subagent is a turn artifact of
/// its parent, not a session: hidden from the timeline, reachable only from the
/// row that spawned it.
struct SessionSubagent: Codable, Hashable, Sendable, Identifiable {
    let sessionId: String
    let provider: String
    let parentToolCallId: String?
    let runId: String?
    let startedAt: String?
    let lastActivityAt: String?
    let endedAt: String?
    let userMessages: Int
    let assistantMessages: Int
    let toolCalls: Int
    let title: String?
    let firstUserMessagePreview: String?
    let lastVisibleTextPreview: String?

    var id: String { sessionId }
}

struct SessionSubagentsResponse: Codable, Sendable {
    let sessionId: String
    let children: [SessionSubagent]
}

/// Binding worker transcripts to the tool call that spawned them. Mirrors
/// `web/src/lib/sessionWorkspace/subagents.ts`; the two must agree, because the
/// same session is read on both surfaces.
///
/// Two shapes, two kinds of provider-supplied evidence. A Task/Agent subagent
/// names its parent tool call in a sidecar the engine reads at parse time. A
/// Workflow subagent knows only its run id, and the one place a run is tied to a
/// tool call is the parent's own tool result, which names the transcript
/// directory.
///
/// Fail closed on both: an unmatched run leaves its children unbound and the row
/// renders as an ordinary tool call. Never "nearest Workflow call".
enum Subagents {
    /// Run ids named by one tool result, in the order they appear.
    static func runIds(inToolOutput output: String?) -> [String] {
        guard let output, output.contains("subagents") else { return [] }
        let pattern = #"Transcript dir:\s*\S*?[/\\]subagents[/\\]workflows[/\\]([A-Za-z0-9_-]+)"#
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return [] }
        var found: [String] = []
        let range = NSRange(output.startIndex..<output.endIndex, in: output)
        for match in regex.matches(in: output, range: range) {
            guard match.numberOfRanges > 1, let captured = Range(match.range(at: 1), in: output) else { continue }
            let value = String(output[captured])
            if !value.isEmpty && !found.contains(value) { found.append(value) }
        }
        return found
    }

    /// Children spawned by one tool call: those naming it directly, plus every
    /// member of a run that call launched.
    static func children(
        from children: [SessionSubagent],
        toolCallId: String?,
        toolOutputText: String?
    ) -> [SessionSubagent] {
        guard !children.isEmpty else { return [] }
        let runIds = runIds(inToolOutput: toolOutputText)
        let matched = children.filter { child in
            if let toolCallId, !toolCallId.isEmpty, child.parentToolCallId == toolCallId { return true }
            guard let runId = child.runId else { return false }
            return runIds.contains(runId)
        }
        return matched.sorted { ($0.startedAt ?? "") < ($1.startedAt ?? "") }
    }

    /// "22 agents · 4m12s" — the shape of the work, before anyone expands it.
    static func summary(_ children: [SessionSubagent]) -> String {
        guard !children.isEmpty else { return "" }
        let label = children.count == 1 ? "1 agent" : "\(children.count) agents"
        let starts = children.compactMap { $0.startedAt.flatMap(parseTimestamp) }
        let ends = children.compactMap { $0.endedAt.flatMap(parseTimestamp) }
        guard !starts.isEmpty, ends.count == children.count,
              let first = starts.min(), let last = ends.max() else { return label }
        let span = last.timeIntervalSince(first)
        guard span > 0 else { return label }
        let seconds = Int(span.rounded())
        let duration = seconds < 60 ? "\(seconds)s" : "\(seconds / 60)m\(String(format: "%02d", seconds % 60))s"
        return "\(label) · \(duration)"
    }

    /// One child's line: its own title, or the prompt it was handed.
    static func label(for child: SessionSubagent) -> String {
        let candidate = (child.title ?? child.firstUserMessagePreview ?? "")
            .replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        if candidate.isEmpty { return String(child.sessionId.prefix(8)) }
        return candidate.count > 80 ? String(candidate.prefix(79)) + "…" : candidate
    }

    private static func parseTimestamp(_ value: String) -> Date? {
        let withFraction = ISO8601DateFormatter()
        withFraction.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return withFraction.date(from: value) ?? ISO8601DateFormatter().date(from: value)
    }
}
